src/goto-instrument/replace_function_bodies.cpp

```diff
@@ -0,0 +1,367 @@
+    end_instruction->make_end_function();
+  }
+};
+
+class assert_false_replace_function_bodiest : public replace_function_bodiest
+{
+protected:
+  void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) override
+  {
+    auto assert_instruction = function.body.add_instruction();
+    assert_instruction->make_assertion(false_exprt());
+    assert_instruction->code = code_assertt(false_exprt());
+    assert_instruction->source_location.set_function(function_name);
+    namespacet ns(symbol_table);
+    assert_instruction->source_location.set_comment(
+      irep_idt(
+        id2string(ID_assertion) + " " +
+        from_expr(ns, "", assert_instruction->guard)));
+    assert_instruction->source_location.set_property_class(ID_assertion);
+    auto end_instruction = function.body.add_instruction();
+    end_instruction->make_end_function();
+  }
+};
+
+class assert_false_then_assume_false_replace_function_bodiest
+  : public replace_function_bodiest
+{
+protected:
+  void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) override
+  {
+    auto assert_instruction = function.body.add_instruction();
+    assert_instruction->make_assertion(false_exprt());
+    assert_instruction->code = code_assertt(false_exprt());
+    assert_instruction->source_location.set_function(function_name);
+    namespacet ns(symbol_table);
+    assert_instruction->source_location.set_comment(
+      irep_idt(
+        id2string(ID_assertion) + " " +
+        from_expr(ns, "", assert_instruction->guard)));
+    assert_instruction->source_location.set_property_class(ID_assertion);
+    auto assume_instruction = function.body.add_instruction();
+    assume_instruction->make_assumption(false_exprt());
+    assume_instruction->code = code_assumet(false_exprt());
+    auto end_instruction = function.body.add_instruction();
+    end_instruction->make_end_function();
+  }
+};
+
+class empty_replace_function_bodiest : public replace_function_bodiest
+{
+protected:
+  void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) override
+  {
+  }
+};
+
+class havoc_replace_function_bodiest : public replace_function_bodiest
+{
+public:
+  havoc_replace_function_bodiest(
+    std::vector<irep_idt> globals_to_havoc,
+    std::regex parameters_to_havoc)
+    : globals_to_havoc(std::move(globals_to_havoc)),
+      parameters_to_havoc(parameters_to_havoc)
+  {
+  }
+
+protected:
+  void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) override
+  {
+    namespacet ns(symbol_table);
+    for(auto const &parameter : function.type.parameters())
+    {
+      if(
+        parameter.type().id() == ID_pointer &&
+        !parameter.type().subtype().get_bool(ID_C_constant) &&
```

Comment:
Not sure, but perhaps this is the same condition as `is_constant_or_has_constant_components` (e.g. is pointer to a pointer to a constant?)

Reply:
I'm not sure; Actually it's also possible that this is the better version; i.e. you can absolutely assign a pointer-to-pointer-to-constant. Then again, structs with constant members usually aren't assignable so I'm not 100% clear on what the correct way to do this here is

Reply:
Consider writing unit tests to figure out - I think it should be clear what is "correct" when you do this